### ISSUE TITLE
Use mold as default linker?

### ISSUE THREAD

**Issue by SUPERCILEX** [rust-lang/rust#94347]
https://github.com/rui314/mold/releases/tag/v1.1

1.1 now supports LTO... is there anything else missing? Or could mold become the default linker?

**Comment by gustav3d** [comment]
Might need to be battle tested a bit first ?.

**Comment by workingjubilee** [comment]
Why?

"It is possible" is not enough. Even "it is faster" is not enough. It is also buggier at the moment, such as with [this miscompilation found on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/.E2.9C.94.20transmute.20usize.20-.3E.20fn.20pointer.20miscompilation.3F/near/271913499)

It has limited cross-platform support:
- https://github.com/rui314/mold/issues/189
- https://github.com/rui314/mold/issues/190

and may contend for resources in more complex multitasking situations (though sometimes LLD does also, being designed by the same hand):
- https://github.com/rui314/mold/issues/117

We have not even finished moving over to lld by default everywhere we can, the process of which would also be somewhat similar to the model for moving over to mold by default anyways, so it's not exactly clear we should be skipping ahead:
- https://github.com/rust-lang/rust/issues/39915
- https://github.com/rust-lang/rust/issues/71515
- https://github.com/rust-lang/rust/issues/71520

And we do not even support using it directly at the moment via an argument like `linker-flavor`.

**Comment by SUPERCILEX** [comment]
Thanks for those details! So step 1 then would be supporting mold as an official linker flavor? Would that be desirable? I do think the performance gains are at least worth making it easy for people to switch if it fits their use case.

**Comment by hkratz** [comment]
Mold is "designed to be a drop-in replacement for the GNU linkers", so a new linker flavor does not seem to be necessary. 
There already is documentation on how to use it with Rust at https://github.com/rui314/mold#how-to-use, I don't think we could make it much simpler without bundling it. `mold -run cargo build` also worked fine for me.

Also, I don't know what the stance on redistributing software licensed under the GNU AGPL v3 is.

**Comment by workingjubilee** [comment]
Ah, I thought [`-Clinker-flavor`](https://doc.rust-lang.org/rustc/codegen-options/index.html#linker-flavor) was for controlling the linker selection in general. Yes, if that's not the intent then there is probably nothing to be done, it seems?

**Comment by Timmmm** [comment]
> Even "it is faster" is not enough.

Why not? Compile time has repeatedly been a top issue for Rust users. Linking is often extremely slow and can dominate build time for incremental compilation.

When it becomes as mature as LLD then "it is faster" definitely seems like a reason to set it as the default.

**Comment by gustav3d** [comment]
Exactly, why not ?.

**Comment by workingjubilee** [comment]
It may still be unwise to use it by default due to lack of jobserver integration. We could certainly consider shipping it along with the compiler, though, if it were not for the legal issues. I have no idea how those will shake out, positive or negative, mind, but I will make the observation has been that legal questions have been answered... slowly. Granted, so will a linker maturing.

If someone sponsored rui314 sufficiently (he is apparently soliciting for sufficient sponsorship so as to allow continuing to work on it and relicensing it under a permissive license), that would stop being an issue, however.

**Comment by shmerl** [comment]
> Mold is "designed to be a drop-in replacement for the GNU linkers", so a new linker flavor does not seem to be necessary.
> There already is documentation on how to use it with Rust at https://github.com/rui314/mold#how-to-use, I don't think we could make it much simpler without bundling it

That's a somewhat confusing way to use mold if that's what you are suggesting:


```
[target.x86_64-unknown-linux-gnu]
linker = "clang"
rustflags = ["-C", "link-arg=-fuse-ld=/path/to/mold"]
```

I.e. why does it have to mention clang?

**Comment by shmerl** [comment]
Just to clarify for anyone who needs it, you can enable usage of mold by creating

`.cargo/config.toml` in your project directory like this (pick your target accordingly):

You still need gcc installed for that as a higher interface around the linker.

```
[target.x86_64-unknown-linux-gnu]
rustflags = ["-C", "link-arg=-fuse-ld=mold"]
```

**Comment by NobodyXu** [comment]
IMHO using mold for debug build makes sense, since it provides faster link time and it's unlikely any people want LTO for their debug build.

**Comment by bjorn3** [comment]
LTO is supported by mold now. And unless you use `-Clinker-plugin-lto` rustc doesn't depend on the linker for LTO anyway.

**Comment by lnicola** [comment]
> It may still be unwise to use it by default due to lack of jobserver integration.

That's also [the case](https://hackmd.io/tFDifkUcSLGoHPBRIl0z8w?view#Missing-jobserver-support) for `lld`, which is currently being stabilized, so it shouldn't cause problems in practice.

**Comment by Kobzol** [comment]
> > It may still be unwise to use it by default due to lack of jobserver integration.
> 
> That's also [the case](https://hackmd.io/tFDifkUcSLGoHPBRIl0z8w?view#Missing-jobserver-support) for `lld`, which is currently being stabilized, so it's not a good argument against `mold` and shouldn't cause problems in practice.

Well, we don't know that yet :) We'll see how the stabilization goes.

In general, I would say that the doors to using mold/wild/other linker as the default in the future are definitely not closed. But first we needed to even figure out a way *how* to switch the default linker, and see what it entails in practice (which we will hopefully find out with https://github.com/rust-lang/rust/pull/140525, but it will take time). LLD has the advantage that it has likely the biggest compatibility with BFD out of the existing "fast" linkers, it is also maintained by LLVM developers, which means that it is shipped with LLVM's source code, which we already depend on, and it will observe the same update cadence as LLVM. We have also already been shipping it in our sysroots for years and it was enabled by default on nightly for almost a year now.

It makes sense to start with LLD to see how such a linker switch goes in practice. Later, if we see that there are benefits to switching to something else, and that the switch hasn't caused (too) many issues in practice, we can start thinking about using a different linker by default on Linux.

**Comment by asyync1024** [comment]
Sorry to ping literally everyone in this thread!

LLD 23 is supposed to have significant performance improvements: https://maskray.me/blog/2026-04-12-recent-lld-elf-performance-improvements

**Approx 34% as of now.**

Further performance work is planned as written in the blog.

This makes switching to `mold` a less feasible option now.

 Also, correctness is always first, then speed. For me, `mold` had linked a 32-bit (on x86_64) Godot Engine binary that crashed on startup about an year ago, it isn't really as reliable as LLD.

**Comment by dralley** [comment]
`wild` linker seems like a more interesting long-term option than `mold`, partly because it's written in Rust, partly because it's even faster than `mold`, and partly because there is an intention and plan around implementing support for incremental linking (but that's a long way off still).

But for now it's good to hear that `lld` is still improving.

**Comment by eclairevoyant** [comment]
That doesn't really discuss correctness, iirc mold was faster because it just did less than lld. If wild is the same then that's just pointless.

Also the linked blogpost is clearly llm slop, I wouldn't trust it.

**Comment by mati865** [comment]
Mold and Wild are faster because they are more parallel.

> Also the linked blogpost is clearly llm slop, I wouldn't trust it.

That's the blog of legendary MaskRay. He is LLD maintainer and likely the most important person of LLD's ELF backend in recent years.

**Comment by eclairevoyant** [comment]
That's lovely, but it's clearly llm slop, claude outputs text exactly like that.
For all we know the code is slop too, then.

For mold I'll find the issue that discusses how it does less than lld and link it here at some point. It's a worse linker overall and I'm not sure how it survived this long.

**Comment by ds84182** [comment]
You can switch to mold yourself if you care that much.

**Comment by asyync1024** [comment]
> wild linker seems like a more interesting long-term option than mold

Agree, though it's even less mature than `mold` as of now, and a linker usually takes at least a few years to mature. `mold` is "production-ready", but it's still not mature enough to be trusted in edge case situations, that's why LLD is still the default.
Not to mention `wild` still has issues in LTO plugin support. [[1]](https://github.com/wild-linker/wild/issues?q=is%3Aissue%20state%3Aopen%20label%3ALTO)

> For mold I'll find the issue that discusses how it does less than lld and link it here at some point.

`mold`'s design.md for reference: https://github.com/rui314/mold/blob/main/docs/design.md

>  If wild is the same then that's just pointless.

If it was "just the same" [[2]](https://davidlattimore.github.io/posts/2025/09/02/rustforge-wild-performance-tricks.html) it wouldn't be improving performance every release while more features are being added, not to say it's **at least 2x faster than `mold`.** [[3]](https://github.com/wild-linker/wild#ryzen-9955hx-16-core-32-thread)

> It's a worse linker overall and I'm not sure how it survived this long.

`mold` started being production ready about 4 years ago [[4]](https://github.com/rui314/mold/releases/tag/v1.0.0) . That's one reason to why LLD "survived" that long. Moreover, LLD was the fastest linker before `mold`. Most of the things holding back LLD are listed in the blog, some more parallelization and per-object work can be improved.

>  For all we know the code is slop too, then.

Hard to say without reading the actual commits that are linked in the post. It does successfully link the binaries on MaskRay's machine though? Not to say, **someone having direct commit access to LLVM sure does have their fair share of credibility.**
NB that LLVM has very high code-quality standards, any sort of code isn't regularly merged without reviews.

**Comment by mati865** [comment]
LLD is not perfect by any means as well in terms of features. For example it lacks linker plugins support. It can preform LTO on LLVM bitcode but not on GCC's one.

So, if your distro does LTO by default for the packaged libraries and some `-sys` crate wants to link them, the only linkers that will work for you are ld.bfd, Mold and Wild.

**Comment by asyync1024** [comment]
Absolutely. But other than GCC LTO support, I would say LLD is very good for it's balance of relatively good speed, and great feature-richness, with great reliability.
